### PR TITLE
Speed up APFloat division by using short division for small divisors.

### LINKED ISSUE #43828: 30% regression in compile time of tuple-stress benchmark
According to [perf.rust-lang.org graphs](https://perf.rust-lang.org/graphs.html?start=2017-07-13T00%3A00%3A00%2B00%3A00&end=2017-08-12T00%3A00%3A00%2B00%3A00&crates=%7B%22list%22%3A%22List%22%2C%22content%22%3A%5B%22tuple-stress%22%5D%7D&phases=%7B%22list%22%3A%22All%22%2C%22content%22%3Anull%7D&group_by=crate&type=time&yaxis=pct) the tuple-stress benchmark has regressed 30% in compile time lately, pointing to https://github.com/rust-lang/rust/pull/43554 as the culprit.

I've downloaded the `nightly-2017-08-05` toolchain and the `nightly-2017-08-06` toolchain where 08-06 has this PR and 08-05 doesn't. The notabl differences in time-passes are:

| pass | before | after |
| -------|------|-------|
| borrow checking | 2.556 | 0.920 |
| const checking | 1.728 | 0.451 |
| translation | 0.798 | 0.682 |
| expansion | 0.168 | 0.097 |

cc @eddyb

### PR DESCRIPTION
**Pull request by eddyb**
Fixes #43828 (hopefully), by not doing long division bit-by-bit for small divisors.

When parsing the ~200,000 decimal float literals in the `tuple-stress` benchmark, this change brings roughly a 5x speed increase (from `0.6s` to `0.12s`), and the hottest instructions are native `div`s.

### CODE REVIEW

**Comment by eddyb** [comment]
@alexcrichton Do you want to review? It think @arielb1 is unavailable for a few days.
Also, any ideas what made the Travis status get stuck? If you look at Travis itself, it's green.

**Comment by alexcrichton** [comment]
I could give this a rubber stamp but unfortunately I don't think I'm equipped to give this an actual review

**Comment by eddyb** [comment]
Hmm. Maybe @est31 or @nagisa could help here?

File: src/librustc_apfloat/ieee.rs [hunk 1]
```diff
@@ -2310,6 +2307,17 @@ mod sig {
         limbs.iter().all(|&l| l == 0)
     }
 
+    /// One, not zero, based LSB. That is, returns 0 for a zeroed significand.
+    pub(super) fn olsb(limbs: &[Limb]) -> usize {
+        for i in 0..limbs.len() {
+            if limbs[i] != 0 {
```
Comment (nagisa) [inline src/librustc_apfloat/ieee.rs]: Why not just iterate the `limbs` slice directly?

Comment (eddyb) [inline src/librustc_apfloat/ieee.rs]: With `enumerate`? That's a transformation that can probably be done in a few places but I didn't do it in the original port.

**Comment by eddyb** [comment]
Oh wow the "rollup" in my previous message confused bors...
...
...
@bors p=7

**Pull request merged**